text/0000-orderer-pruning-and-snapshot.md

```diff
@@ -0,0 +1,275 @@
+OSN ledger snapshot functionality is designed with the same principles as the peer ledger snapshot, in particular it is a full snapshot. We do not implement an incremental snapshot capability.
+
+## Guide: Create snapshot
+
+To create a ledger snapshot use the following REST API (block_number must be supplied as a parameter in the request body):
+
+    POST /v1/mgmt/channels/{channelID}/snapshot
+
+ or the following CLI command:
+
+    $ osnadmin snapshot submitrequest -c <channelID> -b <block_number>
+
+A process of snapshot creation will start immediately if the block number is less than current ledger height. If the block number is in future (>= ledger height) then the snapshot creation will be scheduled and the snapshot will be created when the requested block will appear in the ledger. To check the status of the pending snapshot requests use the following REST API:
+
+    GET /v1/mgmt/channels/{channelID}/snapshotrequests
+
+ or the following CLI command:
+
+    $ osnadmin snapshot listpending -c <channelID>
+
+To cancel a pending snapshot request use the following REST API (block_number must be supplied as a parameter in the request body):
+
+    DELETE /v1/mgmt/channels/{channelID}/snapshotrequests
+
+ or the following CLI command:
+
+    $ osnadmin snapshot cancelrequest -c <channelID> -b <block_number>
+
+## Guide: Join channel from a snapshot
+
+To join a channel from a snapshot use the following REST API (files from the snapshot must be placed in the body, using multipart/form-data content-type):
+
+    POST /v1/mgmt/channels
+
+ or the following CLI command:
+
+    $ osnadmin channel joinbysnapshot -s <snapshot_dir>
+
+To check the status of the "join by snapshot" operation use the following REST API:
+
+    GET /v1/mgmt/channels/{channelID}
+
+ or the following CLI command:
+
+    $ osnadmin channel list -c <channelID>
+
+## Guide: Prune block ledger of an existing channel
+
+To prune an existing channel ledger two approaches may be taken.
+
+First approach: a new snapshot may be created at the desired block number, then an orderer may be removed from the channel and finally the orderer may join the channel using the snapshot.
+This approach has some drawbacks, the main drawbacks are:
+- an interruption in the service of the orderer is required,
+- this works only in the absence of system channel, because "channel joinbysnapshot" works only in the absence of system channel.
+
+Because of the drawbacks, as a part of this RFC we propose to implement a separate **Prune** API which allows to prune the ledger without stopping the service. Essentially this API will create a snapshot and will immediately start using it while deleting all the pruned blocks.
+
+To invoke this new **Prune** feature use the following REST API (block_number must be supplied as a parameter in the request body):
+
+    PATCH /v1/mgmt/channels/{channelID}/prune
+
+ or the following CLI command:
+
+    $ osnadmin channel prune -c <channelID> -b <block_number>
```

Comment:
In my opinion, since `prune` is an add-on operation, it can always be done in a separate RFC. However, If you want to include `prune` in the scope of this RFC, I would suggest to add details about this in the the next section as this requires additional work. In essence, this operation should leave an orderer in a state as if the channel on the orderer  had been bootstrapped from the snapshot. 

The additional work includes deleting the entries from the blockstore index for the blocks prior to the snapshot and  deleting the files (and possibly, splitting one of the files) that contains the prior blocks. Moreover, these operations need to be done in a manner that offers a crash consistent behavior.

On a side note, there are on-going thoughts on implementing the pruning feature for a peer as well. And that would take advantage of this partial work done in this context here.

About actual outcome of the pruning, I would suggest that we could just leave the unused block files in a separate location and let admin decide whether (s)he wants to delete these or archive them somewhere.

Reply:
It is in our best interest to streamline the approval and implementation of the RFC, so I will remove pruning from this RFC and create a separate one, for pruning only.
So this RFC will be limited in scope to snapshot feature only.

Reply:
Done.
Pruning feature removed from the RFC.